part of cam6_3_098:  always pass NDEP from CAM and remove sst specs

mvertens opened the pull request:
This PR addresses issue https://github.com/ESCOMP/CAM/issues/104 and enables CAM to read in ndep (using the CDEPS inline API) from forcing files if it is not computing NDEP prognostically
As a result - CAM will ALWAYS send NDEP to the mediator
Right how, CLM is not using the passed NDEP from the streams - since it is still reading the drv_flds_in for NDEP
Once this PR is accepted the drv_flds_in entry for ndep can be removed and CLM can always accept NDEP from either CAM or DATM.
New XML variables are introduced to specify the stream forcing. This is being done in CTSM as well for all of its CDEPS stream specific variables.

I have run the following test suite on cheyenne with bfb results (other than the FIELDLIST is different for non-WACCM cases since NDEP is now being passed from CAM. 
./create_test --xml-category aux_cam --xml-machine cheyenne --xml-compiler intel --compare cam6_3_088 --baseline-root /glade/p/cesmdata/cseg/cam_baselines/cesm_baselines

NOTE: that CTSM also needs a new PR in order to accept these new fields from CAM rather than use its own internal streams.

Replaces #729 (previous PR which replaced with this one)

Closes https://github.com/ESCOMP/CAM/issues/104

mvertens commented:
@cacraigucar - sorry for the confusion with this PR. This PR has been tested and should be the code to review. I am happy to answer any questions you have and even have a zoom chat if that would be helpful.

cacraigucar commented on cime_config/config_component.xml:
```diff
@@ -348,6 +349,79 @@
+    <type>char</type>
+    <default_value>UNSET</default_value>
+    <values match="last">
+      <value compset="^SSP585_">$DIN_LOC_ROOT/lnd/clm2/ndepdata/fndep_clm_f09_g17.CMIP6-SSP5-8.5-WACCM_1849-2101_monthly_c191007.nc</value>
```
I guess I still have this question - why do we have SSP compsets here?  I don't believe cam_development has any SSP compsets.  Is this to support CESM checkouts?

mvertens replied:
Good question. So CTSM had expected these forcings - but maybe that was only needed for runs with DATM. @billsacks - do you know what the CTSM requirements are. I was simply adding the CTSM forcings back into CAM that it was using. But if CAM development never uses SSP compsets - I would say this should not be supported.

billsacks replied:
I'm not totally sure, but I think this can be removed from CAM. It looks like there were many SSP B compsets in the 2.1 release, but they no longer exist in development, so yes, probably reasonable to remove them here.

cacraigucar replied:
@ekluzek - Here is the PR we discussed yesterday and this is a section which pertains to CTSM

cacraigucar replied:
@mvertens - I believe the consensus is that the SSP references should be removed for this commit

mvertens replied:
@cacraig - I agree - I have removed the SSP reference.

cacraigucar replied:
@mvertens - I do not see them being removed yet, so perhaps you have not pushed the changes yet?

mvertens replied:
@cacraig @nusbaume - the NDEP settings for CAM would be used in fully coupled simulations as well - and thinking about this some more I think we should keep the SSP settings for backwards compatibility. The whole idea is that CAM should be able to obtain forcing data for NDEP that was previously done by both CTSM and POP.

cacraigucar reviewed (commented):
A couple of questions to aid my review

mvertens commented on bld/build-namelist:
```diff
@@ -4023,6 +4040,31 @@ add_default($nl, 'cam_snapshot_before_num');
+  my $var_xml = uc($var_nml);
+  my $var_xml = "CAM_${var_nml}";
+  if ($check_char) {
+      if ($val_nml ne "'$val_opts'"){
```
These new namelist variables can ONLY be changed via xml variables - not in user_nl_cam settings. So this is checking that you did not try to change this variable in user_nl_cam and then issues a warning and aborts. Does that make sense?

cacraigucar replied:
What is confusing me then, is why are they being listed in namelist_defaults.xml and namelist_definition.xml?   As far as I was aware, all variables in the namelists were ones that could be controlled by user_nl_cam.  We have configure options (like -phys cam6) which control the setup and are not changeable.   These variables are feeling like they fit into the configure type of variables.  Perhaps a configure option like "stream_ndep" and then it sets these 5 variables?

A second question about this is, why have these 5 variables exposed at the build-namelist option level?  Up until now, there were only a few options and they were very high level like "ntasks".  

@nusbaume, @jtruesdal,  @fvitt and/or @peverwhee  - Do any of you have more experience with build-namelist and have an opinion on this?

gold2718 replied:
> why are they being listed in namelist_defaults.xml and namelist_definition.xml?

I may be missing something but I think this is a collision between the new (CIME) way of handling namelists and the old-style namelist_d*.xml files used by CAM. In a new `namelist_definition_<comp>.xml` file, this dependency is noted so that this sort of confusion does not exist. For instance, look at the definition of `nx_global` in `<cam_root>/components/cdeps/datm/cime_config/namelist_definition_datm.xml`.

We will (eventually) grow out of this with the new infrastructure.

cacraigucar replied:
@gold2718 - Thank you for the clarification.

cacraigucar replied:
Leaving this unresolved until @nusbaume reviews, so he understands the requirements

mvertens replied:
@nusbaume - if I am understanding your question correctly the idea of putting this in env_run.xml was to coordinate this with datm - so that both datm and cam could read the same xml variables for nitrogen deposition forcing. So since CAM is using streams via CDEPS shared code - it helps if the streams are defined the same way in both. Does that make sense?

brian-eaton replied:
@mvertens, I understand that you want the new namelist vars, stream_ndep_*, to only be settable via the corresponding case variables, CAM_STREAM_NDEP_*.  This is similar to the namelist variables that are set in config_component.xml in the CAM_NAMELIST_OPTS case variable.  My suggestion is that rather than passing the variables as separate arguments to build-namelist, they could be added to the atmexp namelist group that's created in buildnml and passed to build-namelist using its -namelist argument.  That would give these variables the same level of precedence that the CAM_NAMELIST_OPTS variables have, i.e., they can only be changed with xmlchange commands.  Doing it this way would require no changes to the build-namelist script.

brian-eaton replied:
Sorry about my formatting issue in the previous comment.  Trying again...  I meant stream_ndep_\* and CAM_STREAM_NDEP_\*.  Need to remember to use preview:)

brian-eaton replied:
@cacraigucar, @gold2718 - I thought it might be worth a quick note here addressing Cheryl's question about why the new namelist vars are in the namelist_defaults.xml and namelist_definition.xml files.  Since these variables are being set by config_component.xml so the values can be shared by DATM and CAM, then assuming config_component.xml is taking responsibility to always set their values, the variables do not need to be set in namelist_defaults.xml.  However, they still must be in the namelist_definition.xml file in order for build-namelist to put them in the atm_in file.  The definition file contains metadata about the namelist group that the variables belong to.  And in the case of varibles that provide filepaths, it also contains the metadata that tells build-namelist to put those filepaths in the cam.input_data_list file.

cacraigucar commented on bld/namelist_files/use_cases/aquaplanet_cam3.xml:
```diff
@@ -57,9 +57,6 @@
 <!-- shut off prescribed aerosols -->
 <prescribed_aero_file    ></prescribed_aero_file>
 
-<!-- Analytic sst forcing type 1 -->
-<aqua_planet              >.true. </aqua_planet>
-
```
Why is `aqua_planet `removed for the use_cases for cam3 and cam4, but not for cam5, cam6 and rce_cam6 files?  Should they have been removed?

mvertens replied:
@cacraig - you are right. @gold2718 @nusbaume @peverwhee - there is a real problem with how the aquaplanet is treated in cam with nuopc. The aquaplanet flag is coming in from the nuopc driver via a NOUPC attribute. This is a run time setting. However, it seems that CAM in independently setting the aquaplanet config switch and basing other namelist settings based on that. It also seems that this logic is still using the old assumption that CAM had an internal docn functionality. It would be good to have a talk to determine how we want to do this consistently. I think its just fortunate that things seem to be working.

cacraigucar replied:
Leaving this unresolved as well until @nusbaume does his review, so he sees the comment, but I consider this resolved from my viewpoint

mvertens replied:
@jesse - I just created a CAM issue and assigned it to myself.

nusbaume replied:
@mvertens It doesn't look like the new issue in ESCOMP/CAM has shown up on my end.  Are you able to find it?

mvertens replied:
@nusbaume - the issue should be there now. Thanks for pointing this out.

cacraigucar commented on src/cpl/nuopc/atm_import_export.F90:
```diff
@@ -1056,19 +1070,33 @@ subroutine export_fields( gcomp, cam_out, rc)
+       ! expects units of kgN/m2/sec
+       scale_ndep = .001_r8
+    else
+       ! If waccm computes ndep, then its in of kgN/m2/s - and the mediator expects
```
Add the word `units` to `then its in units of`

mvertens replied:
Done.

cacraigucar commented on src/cpl/nuopc/atm_stream_ndep.F90:
```diff
@@ -0,0 +1,247 @@
+       end if
+       close(nu_nml)
+    endif
+    call shr_mpi_bcast(stream_ndep_mesh_filename , mpicom)
+    call shr_mpi_bcast(stream_ndep_data_filename , mpicom)
+    call shr_mpi_bcast(stream_ndep_year_first    , mpicom)
+    call shr_mpi_bcast(stream_ndep_year_last     , mpicom)
+    call shr_mpi_bcast(stream_ndep_year_align    , mpicom)
```
Typically CAM code uses mpi_bcast.  We do not have any calls using shr_mpi_bcast.  We also include the error code return and then check the error and call endrun if an error is encountered.  See src/physics/cam/clubb_intr.F90 for an example.

mvertens commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -3138,96 +3138,31 @@
+
+<stream_year_first_ndep cam_pres_ndep="clim_2000">2010</stream_year_first_ndep>
+<stream_year_last_ndep  cam_pres_ndep="clim_2000">2010</stream_year_last_ndep>
+<model_year_align_ndep  cam_pres_ndep="clim_2000">1</model_year_align_ndep>
```
@cacraigucar - good catch - I fixed it.

mvertens commented:
@cacraigucar - I believe that I have addressed the changes you requested and that this PR is now ready for another review.

cacraigucar commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -3138,96 +3138,31 @@
+<stream_datafile_ndep   cam_pres_ndep="SSP1-4.5" >lnd/clm2/ndepdata/lnd/clm2/fndepdata/fndep_clm_f09_g17.CMIP6-SSP2-4.5-WACCM_1849-2101_monthly_c191007.nc</stream_datafile_ndep>
+<stream_datafile_ndep   cam_pres_ndep="SSP1-7.0" >lnd/clm2/ndepdata/fndep_clm_SSP370_b.e21.BWSSP370cmip6.f09_g17.CMIP6-SSP3-7.0-WACCM.002_1849-2101_monthly_0.9x1.25_c211216.nc</stream_datafile_ndep>
+<stream_datafile_ndep   cam_pres_ndep="SSP1-8.50">lnd/clm2/ndepdata/fndep_clm_f09_g17.CMIP6-SSP5-8.5-WACCM_1849-2101_monthly_c191007.nc</stream_datafile_ndep>
+
+<stream_year_first_ndep cam_pres_ndep="clim_1850">1850</stream_year_first_ndep>
```
@mvertens - Do we need settings for `hist` as well?

mvertens replied:
I think we should retain the settings for hist - since its needed for transitory simulations. I have left it there.

cacraigucar replied:
@mvertens - I was actually asking if we should have settings for `cam_pres_ndep="hist"`?  At line 3150, you are assigning a `stream_datafile_ndep` file for `"hist".`  Do we need the corresponding `stream_year_first_ndep, stream_year_last_ndep and model_year_align_ndep` for `"hist"`?

mvertens replied:
@cacraig, @nusbaume - it turns out that cam_pres_ndep is not used. Sorry for not catching this earlier. All settings for the CAM namelist for NDEP come from the env_run.xml file and those are set correctly. I'm so sorry for all the delay in this PR. We have been traveling and moving non-stop for the last several weeks and I have not had a chance to actually sit down and think a little until now. I have tested and verified that for a HIST setting - the start and stop NDEP times are being set correctly. There is still SSP settings in config_component.xml and for now I see no problem in leaving them there for backwards compatibility.

cacraigucar commented on src/cpl/nuopc/atm_stream_ndep.F90:
```diff
@@ -0,0 +1,253 @@
+    character(*), parameter :: subName = "('stream_ndep_init')"
+    !-----------------------------------------------------------------------
+
+    rc = ESMF_SUCCESS
+
```
@mvertens - Testing discovered this misplaced assignment (it's too high)

mvertens replied:
@cacraigucar -fixed. Although by testing on cheyenne di  not catch this. Very odd.

fvitt reviewed (commented):
If we are always passing 2 nitrogen deposition fields to the mediator via `Faxa_ndep` it seems we can get rid of `Faxa_nhx` and `Faxa_noy`.  Furthermore, I don't see the need for `ndep` namelist settings in `drv_fld_in`.  Using `ndep_list` setting in `drv_flds_in` is a bit of a convoluted way of determining if chemistry will provide the N dep fluxes or if they need to be read from file.

mvertens commented on cime_config/buildnml:
```diff
@@ -169,10 +175,17 @@ def buildnml(case, caseroot, compname):
+        buildnl_opts += ["-stream_ndep_year_align"    , stream_ndep_year_align]
+        buildnl_opts += ["-stream_ndep_data_filename" , stream_ndep_data_filename]
+        buildnl_opts += ["-stream_ndep_mesh_filename" , stream_ndep_mesh_filename]
+
```
@fvitt - you still need the drv_flds_in to determine if waccm will be sending the data or not. The new logic is in atm_import_export.F90:  

     if (ndep_nflds > 0) then
       ! The following is when WACCM computes ndep
       call set_active_Faxa_nhx(.true.)
       call set_active_Faxa_noy(.true.)
    else
       ! The following is used for reading in stream data
       call set_active_Faxa_nhx(.false.)
       call set_active_Faxa_noy(.false.)
    end if
    ! Assume that 2 fields are always sent as part of Faxa_ndep
    call fldlist_add(fldsFrAtm_num, fldsFrAtm, 'Faxa_ndep', ungridded_lbound=1, ungridded_ubound=2)

I agree with you that this is not optimal and we should remove the need for drv_flds_in. However, CTSM is currently ignoring the NDEP sent to it from CAM in the case where ndep_nflds=0. There was a PR I did that removed the internal read of NDEP in CTSM - but unfortunately I did not restrict it to just NDEP and so it is currently on hold. My hope is that CAM can independently always send NDEP and that when the other components start using it and removing any internal read they have - then we can come up with a better mechanism that drv_flds_in for NDEP. Does that make sense?

mvertens replied:
@fvitt - yes datm is sending the prescribed Ndep fluxes to the mediator via Faxa_ndep. But it does not have to be that name for CAM. There are aliases that can be set up in the mediator in fd_cesm.yaml.
My method of determining if ndep should use streams was only an extension of what was there already. If there are more straightforward ways - please described them in an issue. The method I have implemented should be totally backwards compatible with what is already there. I think more straightforward implementations should be done in a separate PR at this point.

fvitt replied:
Sorry if I am being dense. 

In the clm code I see:
```
    ! from atm - nitrogen deposition
    call shr_ndep_readnl("drv_flds_in", ndep_nflds)
    if (ndep_nflds > 0) then
       call fldlist_add(fldsToLnd_num, fldsToLnd, Faxa_ndep, ungridded_lbound=1, ungridded_ubound=ndep_nflds)
       ! This sets a variable in clm_varctl
       ndep_from_cpl = .true.
    end if

```
So clm is going to receive the Faxa_ndep flds only if `ndep_nflds>0`.  Correct?

Meanwhile you are assuming `ndep_nflds>0` when the N dep fluxes are prognosed and do not need to be read from the stream input files.  And the N dep fluxes from CAM will always be sent to the mediator via Faxa_ndep, regardless if they are from the stream inputs or prognosed.

Are there some proposed changes to CLM that I am missing?    It seems that CLM will need to always look for the N dep fluxes in Faxa_ndep independent of the ndep_nflds setting.  Otherwise, CLM will receive the N dep fluxes only if they are prognosed in CAM.

The `Faxa_noy` and `Faxa_nhx` fields should be removed from the system.  They just add to the confusion in my opinion.   I agree things can be cleaned up in another PR which no one will have time to work on...

fvitt replied:
@mvertens 
Thanks for your comments here.  This is very helpful.  Now I am curious what testing has been done for the configurations where CAM is reading and sending the prescribed N dep stream datasets.  Maybe this sort of testing will occur when CLM no long reads the N dep fields internally.

billsacks replied:
@mvertens @fvitt (also @ekluzek @wwieder @adamrher and @olyson since you have been involved in discussions around this) -- I'm trying to make sure I understand what the state of NDEP coupling will be once this PR is merged. I have been confused about whether this PR by itself is sufficient for CTSM to start getting NDEP from ATM, or whether other changes are needed. My understanding from reading the above comments is:
- Currently, CTSM relies on drv_flds_in to determine whether it's getting NDEP from ATM or internally
- After this PR is merged, both CAM and DATM will always provide NDEP
- **But (key point, if I'm understanding this correctly) this PR does NOT have CAM set drv_flds_in for NDEP in the new scenario where CAM is reading NDEP from file. Is that correct?**
- **So (if the above point is correct), after this PR is merged, for CTSM to actually receive NDEP from ATM in all cases, we will need to remove the CTSM logic that only receives NDEP from ATM dependent on drv_flds_in.**
- I don't understand the discussion of Faxa_noy and Faxa_nhx: will anything be needed on the CTSM side based on that discussion? Or, maybe(?) an equivalent question: does CTSM get any information from drv_flds_in that is important other than the yes/no question of whether it's getting NDEP from ATM?

mvertens replied:
@billsacks -
- CAM still sets drv_flds_in for NDEP - since it uses that information to determine if CAM itself needs to read in NDEP from a stream file or use a prognostic calculation. That has not changed. @fvitt pointed out that there are better ways to do this in CAM and this has been deferred to a future PR.
- After this PR is merged it will be up to CTSM to remove the CTSM logic. I did that already in the PR that needs to be reimplemented at this point. CTSM should never need to look into drv_flds_in again for getting NDEP - since it will always be passed in.
- The key point that Francis noted that is the sum of Faxa_noy and Faxa_nhx is all that needs to be passed and not each individual field. All components just sum things up after they get the fields - so sending each field separately is not needed. If CAM and DATM only send the sum then this needs to be a change that is coordinated with multiple components.
Does this help help clarify your questions?

mvertens replied:
@fvitt - in regards to your question as to what testing has been done for the configurations where CAM is reading and sending the prescribed N dep stream datasets - I did careful testing and examined the differences in CTSM history files between the internal read of NDEP in CTSM and what it got when the data came from CAM. Differences were only seen after the 8th place. This should probably be confirmed again.

billsacks replied:
@mvertens - thanks. Your responses help, but to make sure I understand this correctly: Is it correct that, even after this PR, in the case where CAM is reading NDEP from a stream, NDEP will **not** be present in the drv_flds_in file?

mvertens replied:
@billsacks - sorry for not being clear. There CAM does not create an  ndep namelist entry if it needs to read ndep from streams. This is exactly the same behavior as before. So the call to shr_ndep_readnl will return a value of ndep_nflds of 0 - from both CAM and CTSM. When CAM sees the ndep_nflds as 0 - it will read it in from a stream. When CTSM currently sees the ndep_nflds as 0 it will also read it from a stream otherwise if ndep_nflds > 0 it will use the data sent from the mediator (which is the data that CAM sends the mediator). This can clearly all be simplified once CTSM can always just use the data from CAM. Is this clear now?

billsacks replied:
@mvertens - yes, that helps clarify things (and is consistent with what I thought but wasn't sure about). Thank you!

mvertens commented on src/cpl/nuopc/atm_import_export.F90:
```diff
@@ -1056,19 +1070,33 @@ subroutine export_fields( gcomp, cam_out, rc)
+       if (ChkErr(rc,__LINE__,u_FILE_u)) return
+       ! NDEP read from forcing is expected to be in units of gN/m2/sec - but the mediator
+       ! expects units of kgN/m2/sec
+       scale_ndep = .001_r8
```
@fvitt - the units are being checked in atm_stream_ndep (and the model aborts if the units are wrong). Hence the hard-wiring in atm_import_export.F90

    ! Now check to make sure they are correct
    if (.not. trim(ndepunits) == "g(N)/m2/s"  )then
       call endrun(' ERROR in units for nitrogen deposition equal to: '//trim(ndepunits)//" not units expected"// &
            errMsg(sourcefile, __LINE__))
    end if

mvertens replied:
@fvitt - are you satisfied with my response?

nusbaume commented on src/cpl/nuopc/atm_stream_ndep.F90:
```diff
@@ -0,0 +1,253 @@
+       call ESMF_Finalize(endflag=ESMF_END_ABORT)
+    end if
+
+    ! Get pointer for stream data that is time and spatially interpolate to model time and grid
```
grammar fix (`interpolate` -> `interpolated`).

mvertens replied:
Done.

nusbaume reviewed (approved):
Looks good to me, assuming Cheryl and Francis's concerns are resolved.  Otherwise I mostly just have some small typo fixes and a few questions.

mvertens commented on bld/build-namelist:
```diff
@@ -4023,6 +4039,31 @@ add_default($nl, 'cam_snapshot_before_num');
 add_default($nl, 'cam_snapshot_after_num');
 check_snapshot_settings();
 
+# ndep stream forching
```
Done.

mvertens commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -7192,6 +7192,54 @@ List of nitrogen deposition fluxes to be sent from CAM to surface models.
+
+<entry id="stream_ndep_mesh_filename" type="char*256" category="Nitrogen Deposition"
+       group="ndep_stream_nl" valid_values="" >
+NDEP mesh file corresponding to sream_ndep_data_filename.
```
These are no longer present in namelist_definition.xml (see my previous explanation).

mvertens commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -7192,6 +7192,54 @@ List of nitrogen deposition fluxes to be sent from CAM to surface models.
+Set by case xml variable CAM_STREAM_NDEP_MESH_FILENAME.
+</entry>
+
+<entry id="stream_ndep_mapalgo" type="char*256" category="Nitrogen Deposition"
+       group="ndep_stream_nl" valid_values="" >
+NDEP spatial mapping to go from forcing data resolution to model resolution.
+Set by case xml variable CAM_STREAM_NDEP_MAPALGO.
+</entry>
+
+<entry id="stream_ndep_tintalgo" type="char*256" category="Nitrogen Deposition"
+       group="ndep_stream_nl" valid_values="" >
+NDEP time interpolation to go from forcing data times to model times.
+Set by case xml variable CAM_STREAM_NDEP_TINTALGO.
+</entry>
+
+<entry id="stream_ndep_taxmode" type="char*256" category="Nitrogen Deposition"
+       group="ndep_stream_nl" valid_values="" >
+NDEP time extrapolation mode.
+Set by case xml variable CAM_STREAM_DEP_TAXMODE.
+</entry>
```
These are no longer needed in namelist_definition.xml and I have removed them and explained things above.

nusbaume replied:
@mvertens It looks like this code still exists on the PR branch.  Do you just need to push the new commit, or is Github being weird?

cacraigucar replied:
@mvertens  - I agree with @nusbaume that `stream_ndep_mapalgo`, `stream_ndep_tintalgo` and `stream_ndep_taxmode` are still in the `namelist_defintion.xml` file.  I just pulled all of the updates to a clone to make sure.

mvertens replied:
Sorry for the confusion. Its not github. Its me and trying to do this in short sporadic bursts. I have now pushed these changes back.

mvertens commented on src/cpl/nuopc/atm_stream_ndep.F90:
```diff
@@ -0,0 +1,253 @@
+    if (ierr /= 0) call endrun(trim(subname)//": FATAL: mpi_bcast: stream_ndep_data_filename")
+    call mpi_bcast(stream_ndep_year_first, 1, mpi_integer, 0, mpicom, ierr)
+    if (ierr /= 0) call endrun(trim(subname)//": FATAL: mpi_bcast: stream_ndep_year_first")
+    call mpi_bcast(stream_ndep_year_last, 1, mpi_integer, 0, mpicom, ierr)
+    if (ierr /= 0) call endrun(trim(subname)//": FATAL: mpi_bcast: stream_ndep_year_first")
+    call mpi_bcast(stream_ndep_year_align, 1, mpi_integer, 0, mpicom, ierr)
+    if (ierr /= 0) call endrun(trim(subname)//": FATAL: mpi_bcast: stream_ndep_year_first")
```
Done. Thanks for catching this.

mvertens commented:
@nusbaume @cacraigucar @fvitt - I have addressed (hopefully) the issues that were brought up. I am happy to iterate on this in case I missed something. Sorry for the confusion regarding the namelist_definition.xml file. I have removed the relevant section with ndep settings that is no longer needed.

nusbaume commented:
@mvertens Thanks for resolving our concerns, and we totally understand the delay, given what we hear from Steve about how crazy your lives are right now.

Anyways, it does look like there are some changes that appear to be missing, at least from what I can see on Github?  Please let us know (whenever you get the chance) if there was just a forgotten push or two, or if everything looks good on your end.  Thanks!

cacraigucar reviewed (approved):
Thanks @mvertens.  All of my concerns have been addressed.  I bet y'all are glad you are settling into a more permanent location!

mvertens commented:
@cacraigucar - thanks! And yes - we are very happy right now.

nusbaume commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -9039,4 +9045,36 @@ If TRUE perform temp tendency scaling before send to fv3 dynamics
+
+<entry id="stream_datafile_ndep" type="char*256" category="ndep_stream"
+       group="ndep_stream_nml" valid_values="" input_pathname="abs" >
+Stream filename(s) of for Nitrogen Deposition data
```
Grammar change here (`of for` -> `for`).

mvertens replied:
Thanks for catching this. I've put in the fix.

nusbaume reviewed (approved):
Everything looks great now, thanks @mvertens!  I found one minor text change request, but otherwise this PR is good to go from my end.

gold2718 reviewed (approved):
Looks good to me!

billsacks commented:
I think I am only listed as a reviewer because I left review comments. I can't see how to remove myself from the list given that I have already left comments. I have no need to review this. I could mark it as approved if that would help your process, though I'm hesitant about setting the precedent of approving something that I haven't really reviewed at all.

cacraigucar commented:
@billsacks - I am fine leaving your reviewer status as it is.  It indicates to the right that you left comments and I agree that I would not want you to just check the approve review box without actually reviewing it.  I too was unable to remove you from the reviewer list.  I do review folks comments to see whether I should ask for a review or not from them, and I had decided I didn't need to get one from you. @brian-eaton should be moving this PR along shortly.
